akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala

```diff
@@ -56,30 +72,48 @@ private[remote] final class TopHeavyHitters[T](val max: Int) {
    */
   // TODO possibly can be optimised further? (there is a benchmark)
   def update(item: T, count: Long): Boolean =
-    isHeavy(count) && { // O(1) terminate execution ASAP if known to not be a heavy hitter anyway 
+    isHeavy(count) && { // O(1) terminate execution ASAP if known to not be a heavy hitter anyway
       val hashCode = new HashCodeVal(item.hashCode()) // avoid re-calculating hashCode
-      (findHashIdx(0, hashCode): @switch) match { // worst case O(n), can't really bin search here since indexes are kept in synch with other arrays hmm...  
+      val startIndex = hashCode.get & mask
+      (findHashIdx(startIndex, hashCode): @switch) match { // worst case O(n), common O(1 + alpha), can't really bin search here since indexes are kept in synch with other arrays hmm...
         case -1 ⇒
           // not previously heavy hitter
-          insertKnownNewHeavy(hashCode, item, count) // O(1) + rarely O(n) if needs to update lowest hitter
-
+          insertKnownNewHeavy(hashCode, item, count) // O(log n + alpha)
+          true
         case potentialIndexGuess ⇒
           // the found index could be one of many which hash to the same value (we're using open-addressing),
-          // so it is only used as hint for the replace call. If the value matches, we're good, if not we need to search from here onwards.  
+          // so it is only used as hint for the replace call. If the value matches, we're good, if not we need to search from here onwards.
           val actualIdx = findItemIdx(potentialIndexGuess, hashCode, item)
 
-          if (actualIdx == -1) insertKnownNewHeavy(hashCode, item, count) // O(1) + O(n), we simply replace the current lowest heavy hitter
-          else replaceExistingHeavyHitter(actualIdx, hashCode, item, count) // usually O(1), worst case O(n) if we need to scan due to hash conflicts
+          if (actualIdx == -1) {
+            insertKnownNewHeavy(hashCode, item, count) // O(1 + log n), we simply replace the current lowest heavy hitter
+            true
+          } else replaceExistingHeavyHitter(actualIdx, hashCode, item, count) // usually O(1), worst case O(n) if we need to scan due to hash conflicts
```

Comment:
Still, but it could be improved by performing more hash-array maintaining on remove entry. Then it could relaxed to O(alpha).